### ISSUE TITLE
Camel Plugin for Che 7

### ISSUE THREAD

**Issue by slemeur** [eclipse-che/che#12584]
# Camel Plugin for Che 7
## Goals
 * Have features from [VS Code Camel Tooling](https://marketplace.visualstudio.com/items?itemName=camel-tooling.vscode-apache-camel) available in Che 7
   
   * for Camel XML DSL
   * for Camel Java DSL
 
 ## Organization of the work
 Che team will work to implement/fix what is needed to have VS Code Camel extension working. They will provide documentation to Fuse team to integrate the VS Code extension in plugin registry and in stack.
 
 ## Sub-tasks
 **Symbiotic LS running in parallel**
 The Camel tooling is relying on other Language Server to provide its capabilities: XML and Java. Che must be providing support to allow multiple Language Servers, running in parallel.
 
 * [x]  Multiple LS running in parallel
    * [x] Support diagnostic of 2 LS on same file https://github.com/eclipse/che/issues/13154 https://github.com/theia-ide/theia/issues/4929
    * [x] **Support xml, Java and Camel extension altogether seems not working, Java extension is KO https://github.com/eclipse/che/issues/13146**
 
 **XML Plugin in Che**
 The Camel tooling is relying exhaustively on XML. As such, Che should be providing XML tooling, which will work along with the VSCode Camel Extension.
 
 * [x] Support VS Code XML Tooling in Theia [theia-ide/theia#3317](https://github.com/theia-ide/theia/issues/3317)
 * [x]  Add Theia Plugin APIs to support VS Code XML Tooling #11775
 * [x]  Cannot read property 'editor.autoClosingBrackets' of undefined #12536
 
 _Will fix:_
 
 * [x]  Provide an xml editor with xsd completion and validation #6925
 * [x] Che 7 plugin: vscode-xml https://github.com/eclipse/che/issues/12437
 
 **Java Plugin in Che**
 In order for the Camel extension to be working in Che, it requires the Java LS. It also requires to be running as a Che Plugin (running in its own sidecar container).
 
 * [x]  Che Java Plugin: #12436
 * [x]  Theia Java Plugin: [theia-ide/theia#3778](https://github.com/theia-ide/theia/issues/3778)
 

**Camel Plugin in Che**

- [x] Avoid case of diagnostic with empty message reason https://github.com/camel-tooling/camel-language-server/issues/225
- [ ] Avoid log for undocumented protocol messages ( $/setTraceNotification ) https://github.com/camel-tooling/camel-language-server/issues/226
- [x] "Request completionItem/resolve failed." fro Camel VS Code extension https://github.com/theia-ide/theia/issues/4856
- [x] "Error: Request 'updateContent' failed" in docker output when there is a validation error on a Camel file https://github.com/theia-ide/theia/issues/4857
- [x] "Find all references" is not available https://github.com/eclipse/che/issues/13183

 **Camel specific features working with XML**
 
 * Have Camel specific features and classic XML support working with XML in Theia in the same editor
 * Verify that Camel specific features and classic XML support working with XML in Che in the same editor
 
 **Camel specific features working with Java**
 
 * Have Camel specific features and classic Java support working with Java in Theia in the same editor
 * Have Camel specific features and classic Java support working with Java in Che in the same editor
   
   * [x]  with a proper Java project (can use [a Camel SpringBoot example](https://github.com/jboss-fuse/fuse-rest-http-booster))
   * [ ]  with a standalone java file out of a project (lower priority)
 * Be able to install successfully in Che the Camel VS Code extension
 
 **Packaging of the Camel VSCode extension in Che**
 
 * [x]  Have Camel VS Code extension in plugin registry https://github.com/eclipse/che-plugin-registry/issues/109
 
 **Che Stacks Integration**
 
 * [x]  Have a Fuse stack using the Camel VS Code extension from the plugin registry with sample projects to be used with https://github.com/eclipse/che/issues/13182
 
 **Documentation**
 
 * [ ]  Publication of a blog post about Camel on Che Blog
 * [ ]  Publication of an how to use Camel in Eclipse Che on the documentation
 
 **QE:Integration Tests**
 
 * [ ]  Have tests for Eclipse Theia integration
 * [ ]  Have test for Eclipse Che integration
 
 ## Notes
 Main "Camel specific features" are described in the description https://marketplace.visualstudio.com/items?itemName=camel-tooling.vscode-apache-camel
 It includes:
 
 * completion of Camel URI
 * Hover on Camel URI
 * Document symbol for Camel Context and Camel route
 * Diagnostic on Camel URI

**Comment by apupier** [comment]
waiting for the rights to edit this epic, I put here issues that I detected with new version of Theia which allows to test Camel plugin and XMl plugin altogether:
- https://github.com/theia-ide/theia/issues/4857
- https://github.com/theia-ide/theia/issues/4856
- https://github.com/theia-ide/theia/issues/4855

**Comment by tsmaeder** [comment]
Running the API usage report (https://github.com/eclipse/che/issues/12996) shows no missing API

**Comment by tsmaeder** [comment]
> How do I build the extension? npm install fails with Error: Cannot find module 'mvn-artifact-download'

> you need to build one time with npm install --ignore-script
https://github.com/camel-tooling/camel-lsp-client-vscode/blob/master/Contributing.md#how-to-start-development

**Comment by apupier** [comment]
created https://github.com/eclipse/che-plugin-registry/issues/109 for the inclusion of Camel extension in Che plugin registry

**Comment by apupier** [comment]
Duplicating to propose an updated version as i don't have edit rights on original comment

# Camel Plugin for Che 7
## Goals
 * Have features from [VS Code Camel Tooling](https://marketplace.visualstudio.com/items?itemName=camel-tooling.vscode-apache-camel) available in Che 7
   
   * for Camel XML DSL
   * for Camel Java DSL
 
 ## Organization of the work
 Che team will work to implement/fix what is needed to have VS Code Camel extension working. They will provide documentation to Fuse team to integrate the VS Code extension in plugin registry and in stack.
 
 ## Sub-tasks
 **Symbiotic LS running in parallel**
 The Camel tooling is relying on other Language Server to provide its capabilities: XML and Java. Che must be providing support to allow multiple Language Servers, running in parrallel.
 
 * [ ]  Multiple LS running in parallel
 
 **XML Plugin in Che**
 The Camel tooling is relying exhaustively on XML. As such, Che should be providing XML tooling, which will work along with the VSCode Camel Extension.
 
 * [x]  [theia-ide/theia#3317](https://github.com/theia-ide/theia/issues/3317)
 * [ ]  #11775
 
 _Will fix:_
 
 * [ ]  Provide an xml editor with xsd completion and validation #6925
 
 **Java Plugin in Che**
 In order for the Camel extension to be working in Che, it requires the Java LS. It also requires to be running as a Che Plugin (running in its own sidecar container).
 
 * [x]  Che Java Plugin: #12436
 * [ ]  Theia Java Plugin: [theia-ide/theia#3778](https://github.com/theia-ide/theia/issues/3778)
 
 **Camel specific features working with XML**
 
 * Have Camel specific features and classic XML support working with XML in Theia in the same editor
   
   * [ ]  Add Theia Plugin APIs to support VS Code XML Tooling #11775
   * [ ]  Cannot read property 'editor.autoClosingBrackets' of undefined #12536
 * Verify that Camel specific features and classic XML support working with XML in Che in the same editor
 
 **Camel specific features working with Java**
 
 * Have Camel specific features and classic Java support working with Java in Theia in the same editor
 * Have Camel specific features and classic Java support working with Java in Che in the same editor
   
   * [ ]  with a proper Java project (can use [a Camel SpringBoot example](https://github.com/jboss-fuse/fuse-rest-http-booster))
   * [ ]  with a standalone java file out of a project (lower priority)
 * Be able to install succesfully in Che the Camel VS Code extension
 
 **Packaging of the Camel VSCode extension in Che**
 
 * [ ]  Have Camel VS Code extension in plugin registry
 
 **Che Stacks Integration**
 
 * [ ]  Have a Fuse stack using the Camel VS Code extension from the plugin registry with sample projects to be used with
 
 **Documentation**
 
 * [ ]  Publication of a blog post about Camel on Che Blog
 * [ ]  Publication of an how to use Camel in Eclipse Che on the documentation
 
 **QE:Integration Tests**
 
 * [ ]  Have tests for Eclipse Theia integration
 * [ ]  Have test for Eclipse Che integration
 
 ## Notes
 Main "Camel specific features" are described in the description https://marketplace.visualstudio.com/items?itemName=camel-tooling.vscode-apache-camel
 It includes:
 
 * completion of Camel URI
 * Hover on Camel URI
 * Document symbol for Camel Context and Camel route
 * Diagnostic on Camel URI

**Comment by slemeur** [comment]
Closing this epic as most of the work has been completed. Missing the documentation on Che docs - https://github.com/eclipse/che/issues/14778

**Comment by apupier** [comment]
another important part is missing non-regression tests
